### PR TITLE
feat: file per table

### PR DESCRIPTION
**Pull request by fdlk**
TODO:
- [x] Process comments
- [x] Update test.http files
- [x] Check terminology

### CODE REVIEW

File: armadillo/src/main/java/org/molgenis/armadillo/command/Commands.java [hunk 1]
```diff
@@ -13,9 +13,11 @@
...
 
-  List<Workspace> listWorkspaces(String bucketName);
+  List<String> listSharedTables();
+
+  CompletableFuture<Void> loadTable(String symbol, String table, String variables);
 
-  CompletableFuture<List<String>> loadWorkspaces(List<String> workspaces);
+  List<Workspace> listWorkspaces(String bucketName);
 
   CompletableFuture<Void> loadUserWorkspace(String bucketName, String objectName);
```
Comment (fdlk) [inline armadillo/src/main/java/org/molgenis/armadillo/command/Commands.java:22]: Make up your mind! is it domain terminology or object storage terminology

File: armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java [hunk 2]
```diff
@@ -103,31 +109,26 @@ protected Void doWithConnection(RConnection connection) {
...
-              String objectName = workspace.substring(index + 1) + ".RData";
-              InputStream inputStream = storageService.load(bucketName, objectName);
-              rExecutorService.loadWorkspace(
-                  connection, new InputStreamResource(inputStream), TABLE_ENV);
-            };
-          }
-        });
+  @Override
+  public List<String> listSharedTables() {
+    var stringStringMultimap = tableService.listTables();
```
Comment (fdlk) [inline armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java:128]: maybe move from tableservice to here?

File: armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java [hunk 3]
```diff
@@ -103,31 +109,26 @@ protected Void doWithConnection(RConnection connection) {
...
-              String bucketName = "shared-" + workspace.substring(0, index);
-              String objectName = workspace.substring(index + 1) + ".RData";
-              InputStream inputStream = storageService.load(bucketName, objectName);
-              rExecutorService.loadWorkspace(
-                  connection, new InputStreamResource(inputStream), TABLE_ENV);
-            };
-          }
-        });
+  @Override
+  public List<String> listSharedTables() {
```
Comment (fdlk) [inline armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java:127]: add docs: this is all tables in all buckets you're allowed to load from

File: armadillo/src/main/java/org/molgenis/armadillo/minio/MinioStorageService.java [hunk 4]
```diff
@@ -102,11 +125,12 @@ public InputStream load(String bucketName, String objectName) {
     }
   }
 
+  //  @PreAuthorize("hasPermission(#bucketName, 'Bucket', 'load')")
```
Comment (fdlk) [inline armadillo/src/main/java/org/molgenis/armadillo/minio/MinioStorageService.java:128]: fix!

File: armadillo/src/test/java/org/molgenis/armadillo/minio/MinioStorageServiceTest.java [hunk 5]
```diff
@@ -80,29 +75,7 @@ void saveThrowsException() throws Exception {
...
-    when(minioClient.bucketExists("user-admin")).thenReturn(true);
-    when(minioClient.listObjects("user-admin")).thenReturn(List.of(itemResult));
-    when(itemResult.get()).thenReturn(item);
-    when(item.objectName()).thenReturn("blah.RData");
-    when(item.lastModified()).thenReturn(Date.from(lastModified));
-    when(item.etag()).thenReturn(workspace.eTag());
-    when(item.objectSize()).thenReturn(workspace.size());
-
-    assertEquals(List.of(workspace), minioStorageService.listWorkspaces("user-admin"));
+    assertEquals(emptyList(), minioStorageService.listObjects("user-admin"));
```
Comment (fdlk) [inline armadillo/src/test/java/org/molgenis/armadillo/minio/MinioStorageServiceTest.java:78]: add permission tests here now

File: r/src/main/java/org/molgenis/r/service/RExecutorServiceImpl.java [hunk 6]
```diff
@@ -71,6 +74,40 @@ public void loadWorkspace(RConnection connection, Resource resource, String envi
...
+                      .map(String::trim)
+                      .collect(toList())
+                      .toArray(new String[] {}));
+          connection.eval(
+              format(
+                  "is.null(base::assign('%s', value={arrow::read_parquet('%s', col_select = %s)}))",
+                  symbol, rFileName, colSelect));
+        }
+      } else {
+        // TODO: .rda?
```
Comment (fdlk) [inline r/src/main/java/org/molgenis/r/service/RExecutorServiceImpl.java:103]: for now: throw exception

File: armadillo/src/main/java/org/molgenis/armadillo/ArmadilloPermissionEvaluator.java [hunk 7]
```diff
@@ -15,13 +16,23 @@
 public class ArmadilloPermissionEvaluator implements PermissionEvaluator {
 
   public static final String WORKSPACE = "Workspace";
```
Comment (fdlk) [inline armadillo/src/main/java/org/molgenis/armadillo/ArmadilloPermissionEvaluator.java:18]: Remove workspace, check only buckets

File: armadillo/src/main/java/org/molgenis/armadillo/ArmadilloPermissionEvaluator.java [hunk 8]
```diff
@@ -51,15 +64,19 @@ public boolean hasPermission(
...
+    String bucketName = objectName.substring(0, objectName.indexOf('/'));
+    return hasBucketPermission(authentication, bucketName, permission);
+  }
+
+  boolean hasBucketPermission(Authentication authentication, String bucketName, String permission) {
     switch (permission) {
       case LOAD:
         {
-          String bucketName = objectName.substring(0, objectName.indexOf('/'));
           return getRoles(authentication)
```
Comment (fdlk) [inline armadillo/src/main/java/org/molgenis/armadillo/ArmadilloPermissionEvaluator.java:75]: also check user buckets

**Review by fdlk — COMMENTED** [review]
From reviewing with @tommydeboer

File: r/src/main/java/org/molgenis/r/service/RExecutorServiceImpl.java [hunk 9]
```diff
@@ -76,6 +79,40 @@ public void loadWorkspace(RConnection connection, Resource resource, String envi
...
+                      .toArray(new String[] {}));
+          connection.eval(
+              format(
+                  "is.null(base::assign('%s', value={arrow::read_parquet('%s', col_select = %s)}))",
+                  symbol, rFileName, colSelect));
+        }
+      } else {
+        throw new UnsupportedOperationException("Can only load .parquet files");
+      }
+      connection.eval(format("base::unlink('%s')", rFileName));
```
Comment (fdlk) [inline r/src/main/java/org/molgenis/r/service/RExecutorServiceImpl.java:110]: Also remove?

File: armadillo/src/main/java/org/molgenis/armadillo/command/Commands.java [hunk 10]
```diff
@@ -13,9 +13,11 @@
 
   CompletableFuture<Void> assign(String symbol, String expression);
 
-  List<Workspace> listWorkspaces(String bucketName);
+  List<String> listSharedTables();
+
+  CompletableFuture<Void> loadTable(String symbol, String table, String variables);
 
-  CompletableFuture<List<String>> loadWorkspaces(List<String> workspaces);
+  List<Workspace> listWorkspaces(String bucketName);
```
[the remark this answers is not in the document]
Comment (fdlk) [reply armadillo/src/main/java/org/molgenis/armadillo/command/Commands.java]: After giving it a bit more thought, I think listWorkspaces is more concise and would prefer the shorter form.

File: armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java (hunk 3 above)
Comment (fdlk) [reply armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java]: now takes project as argument

File: armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java [hunk 11]
```diff
@@ -73,12 +77,30 @@ public CommandsImpl(
...
     lastCommand = command;
-    CompletableFuture<T> result =
-        supplyAsync(() -> session.execute(command::evaluate), executorService);
+    Supplier<T> execution = withCurrentSecurityContext(() -> session.execute(command::evaluate));
+    CompletableFuture<T> result = supplyAsync(execution, executorService);
     command.setExecution(result);
     return result;
   }
 
+  private <V> Supplier<V> withCurrentSecurityContext(Supplier<V> supplier) {
```
Comment (fdlk) [inline armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java:86]: This was a bit tricky. The armadilloService now does the security checks. but the security context is only available on the thread that does the scheduling. So here we decorate a supplier to run in the current security context.

File: armadillo/src/test/java/org/molgenis/armadillo/ArmadilloSessionTest.java [hunk 12]
```diff
@@ -120,22 +122,24 @@ void sessionCleanupTerminatesRunningProcess() throws RserveException, Interrupte
...
+                          sessionIsDestroyed.await();
+                        } catch (InterruptedException ignore) {
+                        }
+                        return new REXPNull();
+                      });
+                });
     executionIsRunning.await();
     armadilloSession.sessionCleanup();
     sessionIsDestroyed.countDown();
+    task.get();
```
Comment (fdlk) [inline armadillo/src/test/java/org/molgenis/armadillo/ArmadilloSessionTest.java:142]: fixes flaky test, if the task is still running, the verify can fail

**Comment by sidohaakma** [comment]
[Implement one file per table principle in Armadillo service](https://trello.com/c/u12p2VfM/403-implement-one-file-per-table-principle-in-armadillo-service)

File: armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java (hunk 11 above)
Comment (sidohaakma) [reply armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java:82]: I did not know you could do this. What are Suppliers and why should you use it?

File: armadillo/src/test/java/org/molgenis/armadillo/DataControllerTest.java [hunk 13]
```diff
@@ -421,71 +412,49 @@ void testAssignAsync() throws Exception {
...
-  @Test
-  void testLoadTibblesForbidden() throws Exception {
-    mockMvc
-        .perform(post("/load-tables?workspace=DIABETES/patient&workspace=GECKO/patient"))
-        .andExpect(status().isForbidden());
-    verifyNoInteractions(commands);
-  }
-
   @ParameterizedTest
   @ValueSource(
```
Comment (sidohaakma) [inline armadillo/src/test/java/org/molgenis/armadillo/DataControllerTest.java:449]: Nice!

File: armadillo/src/test/java/org/molgenis/armadillo/minio/ArmadilloStorageServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.access.AccessDeniedException;
+import org.springframework.security.config.annotation.method.configuration.EnableGlobalMethodSecurity;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
+
+@SpringJUnitConfig
```
Comment (sidohaakma) [inline armadillo/src/test/java/org/molgenis/armadillo/minio/ArmadilloStorageServiceTest.java:28]: Haha nice catch!

File: armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java (hunk 11 above)
Comment (fdlk) [reply armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java:82]: Supplier<V> is the functional interface for a function that has no arguments and returns V

Comment (fdlk) [reply armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java:82]: They are used as a replacement for Callable by the CompletableFuture framework.

Comment (sidohaakma) [reply armadillo/src/main/java/org/molgenis/armadillo/command/impl/CommandsImpl.java:82]: Ah nice!

**Pull request merged**